cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/TopicReplicationFactorAnomalyFinder.java

```diff
@@ -0,0 +1,94 @@
+import static com.linkedin.kafka.cruisecontrol.detector.AnomalyDetectorUtils.ANOMALY_DETECTION_TIME_MS_OBJECT_CONFIG;
+
+
+/**
+ * The class will check whether there are topics having partition(s) with replication factor different than the desired value.
+ */
+public class TopicReplicationFactorAnomalyFinder implements TopicAnomalyFinder {
+  public static final String SELF_HEALING_TARGET_TOPIC_REPLICATION_FACTOR_CONFIG = "self.healing.target.topic.replication.factor";
+  public static final short DEFAULT_SELF_HEALING_TARGET_TOPIC_REPLICATION_FACTOR = 3;
+  public static final String TOPIC_EXCLUDED_FROM_REPLICATION_FACTOR_CHECK = "topic.excluded.from.replication.factor.check";
+  public static final String DEFAULT_TOPIC_EXCLUDED_FROM_REPLICATION_FACTOR_CHECK = "";
+  public static final String TOPIC_REPLICATION_FACTOR_ANOMALY_CLASS_CONFIG = "topic.replication.topic.anomaly.class";
+  public static final Class<?> DEFAULT_TOPIC_REPLICATION_FACTOR_ANOMALY_CLASS = TopicReplicationFactorAnomaly.class;
+  public static final String TOPICS_WITH_BAD_REPLICATION_FACTOR_CONFIG = "topics.with.bad.replication.factor";
+  private KafkaCruiseControl _kafkaCruiseControl;
+  private int _targetReplicationFactor;
+  private Pattern _topicExcludedFromCheck;
+  private Class<?> _topicReplicationTopicAnomalyClass;
+
+  @Override
+  public List<TopicAnomaly> topicAnomalies() {
+    Cluster cluster = _kafkaCruiseControl.kafkaCluster();
+    Set<String> topicsWithBadReplicationFactor = new HashSet<>();
+    for (String topic : cluster.topics()) {
+      if (_topicExcludedFromCheck.matcher(topic).matches()) {
+        continue;
+      }
+      for (PartitionInfo partition : cluster.partitionsForTopic(topic)) {
+        if (partition.replicas().length != _targetReplicationFactor) {
+          topicsWithBadReplicationFactor.add(topic);
+          break;
+        }
+      }
+    }
+    if (!topicsWithBadReplicationFactor.isEmpty()) {
+      return Collections.singletonList(createTopicReplicationFactorAnomaly(topicsWithBadReplicationFactor,
+                                                                           _targetReplicationFactor));
+    }
+    return Collections.emptyList();
+  }
+
+  private TopicAnomaly createTopicReplicationFactorAnomaly(Set<String> topicsWithBadReplicationFactor,
+                                                          int targetReplicationFactor) {
+    Map<String, Object> configs = new HashMap<>(4);
+    configs.put(KAFKA_CRUISE_CONTROL_OBJECT_CONFIG, _kafkaCruiseControl);
+    configs.put(TOPICS_WITH_BAD_REPLICATION_FACTOR_CONFIG, topicsWithBadReplicationFactor);
+    configs.put(SELF_HEALING_TARGET_TOPIC_REPLICATION_FACTOR_CONFIG, targetReplicationFactor);
+    configs.put(ANOMALY_DETECTION_TIME_MS_OBJECT_CONFIG, _kafkaCruiseControl.timeMs());
+    return getConfiguredInstance(_topicReplicationTopicAnomalyClass, TopicReplicationFactorAnomaly.class, configs);
```

Comment:
The issue with ` KafkaCruiseControlConfig#getConfiguredInstance` is that using this method you can only create object of one type of `TopicAnomaly`.

Imagine we are trying to that method, we will define a `TOPIC_ANOMALY_CLASS_CONFIG` config in [AnomalyDetectorConfig](https://github.com/linkedin/cruise-control/blob/48f8f0f90e56869bc5bed19d04983bea8aba7f24/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/constants/AnomalyDetectorConfig.java#L29) , and lets's say we set the value of this config to be `TopicReplicationFactorAnomaly.class`.  

`TopicReplicationFactorAnomalyFinder` will work just fine. 

Then we are trying to add another `TopicAnomalyFinder`, `TopicPartitionSizeAnomalyFinder` and it is supposed to generate `TopicPartitionSizeAnomaly`, this finder is unable to generate anomaly object using `KafkaCruiseControlConfig#getConfiguredInstance`.

Reply:
I think MeticAnomaly has the same issue.

Reply:
Shouldn't the `*AnomalyFinder` of a particular Anomaly Detector always generate the same type of anomalies?
-- i.e. Shouldn't the generated anomaly type have an API to support any anomaly finders? If the anomaly type is missing certain functionalities, I think we should evolve its API. Note that in case of metric anomalies, we have `SlowBrokers` extending `KafkaMetricAnomaly`. Hence, the metric anomaly detector always returns an instance of `KafkaMetricAnomaly` as expected.